basement/Basement/Bits.hs

```diff
@@ -0,0 +1,486 @@
+-- License     : BSD-style
+-- Maintainer  : Haskell Foundation
+-- Stability   : experimental
+-- Portability : portable
+--
+
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Basement.Bits
+    ( Bit
+    , BitOps(..)
+    , PrimBitOps(..)
+
+    , Bits
+    , toBits
+    , allOne
+    ) where
+
+import Basement.Compat.Base
+import Basement.Compat.Natural
+import Basement.Numerical.Additive
+import Basement.Numerical.Subtractive
+import Basement.Numerical.Multiplicative
+import Basement.Types.OffsetSize
+import Basement.Types.Word128 (Word128)
+import qualified Basement.Types.Word128 as Word128
+import Basement.Types.Word256 (Word256)
+import qualified Basement.Types.Word256 as Word256
+import Basement.IntegralConv (wordToInt)
+import Basement.Nat
+
+import qualified Prelude
+import qualified Data.Bits as OldBits
+import Data.Maybe (fromMaybe)
+import Data.Proxy
+import GHC.Base hiding ((.))
+import GHC.Prim
+import GHC.Types
+import GHC.Word
+import GHC.Int
+
+-- | type level tag to represet a bit.
+--
+-- Mainly used for offsets and sizes
+type Bit = Bool
+
+-- | operation over finit bits
+class PrimBitOps bits where
+    -- | get the number of bits in the given object
+    --
+    numberOfBits :: bits -> CountOf Bit
+
+    -- | rotate the given bit set.
+    rotateL :: bits -> CountOf Bit -> bits
+    -- | rotate the given bit set.
+    rotateR :: bits -> CountOf Bit -> bits
+
+    -- | count of number of bit set to 1 in the given bit set.
+    popCount :: bits -> CountOf Bit
+
+    -- | count of the number of leading zeros
+    countLeadingZeros :: bits -> CountOf Bit
+    default countLeadingZeros :: BitOps bits => bits -> CountOf Bit
+    countLeadingZeros n = loop stop azero
+      where
+        stop = numberOfBits n
+        loop idx count
+            | idx == azero = count
+            | isBitSet n (sizeAsOffset idx) = count
+            | otherwise = loop (fromMaybe azero (idx - 1)) (count + 1)
+
+    -- | count of the number of trailing zeros
+    countTrailingZeros :: bits -> CountOf Bit
+    default countTrailingZeros :: BitOps bits => bits -> CountOf Bit
+    countTrailingZeros n = loop azero
+      where
+        stop = numberOfBits n
+        loop count
+            | count == stop = count
+            | isBitSet n (sizeAsOffset count) = count
+            | otherwise = loop (count + 1)
+
+-- | operation over bits
+class BitOps bits where
+    (.&.)     :: bits -> bits -> bits
+    (.|.)     :: bits -> bits -> bits
+    (.^.)     :: bits -> bits -> bits
+    (.<<.)    :: bits -> CountOf Bit -> bits
+    (.>>.)    :: bits -> CountOf Bit -> bits
+    -- | reverse all bits in the argument
+    not_      :: bits -> bits
```

Comment:
`not_` a good name. `bitwiseCompletement` or `bitwiseNegate` make more sense. also this is a finite operation, so it shouldn't be in this class

Reply:
For info, I like `invert` as the phrases like bitwise are a bit redundant in the Bits class and negate/complement are too overloaded.

Reply:
I agree that in the declaration side it's obvious, but on the other hand in the caller, that make it obvious that you're talking about a bitwise operation. I can't really think of any term that is not overloaded here (negate, completement, or invertion)

Reply:
It's the length of the identifier I don't like, since both `bitwise` and `Complement` are long. Perhaps `bitflip`?

Reply:
`bitflip` (or maybe `bitFlip`) sounds fine; I justified the longer name just because unless you're doing bitwise stuff all the time, it's not very common in code.

Reply:
thanks for the feedback. I have made the required changes.